### PR TITLE
Adding new validation for Json, TEXT indexing

### PR DESCRIPTION
**Pull request by icefury71**
- Adding semantic validation for indexing and field config

## Description
Related to #5942 . This PR adds some missing config validation for range, varLength, Json, TEXT, FST indexing config

## Upgrade Notes
Does this PR prevent a zero down-time upgrade?
No
Does this PR fix a zero-downtime upgrade introduced earlier?
No
Does this PR otherwise need attention when creating release notes?
No

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java [hunk 1]
```diff
@@ -474,11 +478,46 @@ private static void validateIndexingConfig(@Nullable IndexingConfig indexingConf
...
+            continue;
+          default:
+            throw new IllegalStateException(
+                "var length dictionary can only be created for columns of type STRING and BYTES. Invalid for column "
+                    + varLenDictCol);
+        }
+      }
+    }
+
+    if (indexingConfig.getJsonIndexColumns() != null) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java:509]: Json index can only be applied to SV string column

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java [hunk 2]
```diff
@@ -488,18 +527,37 @@ private static void validateFieldConfigList(@Nullable List<FieldConfig> fieldCon
...
+      switch (fieldConfig.getIndexType()) {
+        case FST:
+          Preconditions.checkArgument(fieldConfig.getEncodingType() == FieldConfig.EncodingType.DICTIONARY,
+              "FST Index is only enabled on dictionary encoded columns");
+          Preconditions.checkState(
+              fieldConfigColSpec.isSingleValueField() && fieldConfigColSpec.getDataType() == FieldSpec.DataType.STRING,
+              "FST Index is only supported for single value string columns");
+          break;
+        case TEXT:
+          Preconditions.checkArgument(fieldConfig.getEncodingType() == FieldConfig.EncodingType.RAW,
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java:556]: Text index is supported on dictionary encoded SV string columns. Why add this check?

Comment (icefury71) [reply pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java:556]: That's my bad - removing this check.

Comment (icefury71) [reply pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java:556]: My bad - removing this extra check

**Review by siddharthteotia — APPROVED** [review]
LGTM

File: pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java [hunk 3]
```diff
@@ -474,11 +478,48 @@ private static void validateIndexingConfig(@Nullable IndexingConfig indexingConf
...
+    if (indexingConfig.getVarLengthDictionaryColumns() != null) {
+      for (String varLenDictCol : indexingConfig.getVarLengthDictionaryColumns()) {
+        FieldSpec varLenDictFieldSpec = schema.getFieldSpecFor(varLenDictCol);
+        switch (varLenDictFieldSpec.getDataType()) {
+          case STRING:
+          case BYTES:
+            continue;
+          default:
+            throw new IllegalStateException(
+                "var length dictionary can only be created for columns of type STRING and BYTES. Invalid for column "
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java:503]: What about MV columns, do we support var length for MV columns?

Also, s/var length/Var length

File: pinot-core/src/test/java/org/apache/pinot/core/util/TableConfigUtilsTest.java [hunk 4]
```diff
@@ -870,6 +916,69 @@ public void testValidateIndexingConfig() {
...
+    }
+
+    tableConfig = new TableConfigBuilder(TableType.OFFLINE).setTableName(TABLE_NAME).setRangeIndexColumns(columnList)
+        .setNoDictionaryColumns(columnList).
+            build();
+    try {
+      TableConfigUtils.validate(tableConfig, schema);
+      Assert.fail("Should fail for range index defined on non numeric/no-dictionary column");
+    } catch (Exception e) {
+      // Expected
```
Comment (mayankshriv) [inline pinot-core/src/test/java/org/apache/pinot/core/util/TableConfigUtilsTest.java:962]: Swallowing exception?

File: pinot-core/src/test/java/org/apache/pinot/core/util/TableConfigUtilsTest.java [hunk 5]
```diff
@@ -870,6 +916,69 @@ public void testValidateIndexingConfig() {
...
+      // expected
+    }
+
+    tableConfig = new TableConfigBuilder(TableType.OFFLINE).setTableName(TABLE_NAME).setJsonIndexColumns(Arrays.asList("multiValCol")).
+        build();
+    try {
+      TableConfigUtils.validate(tableConfig, schema);
+      Assert.fail("Should fail for Json Index defined on a multi value column");
+    } catch (Exception e) {
+      // expected
```
Comment (mayankshriv) [inline pinot-core/src/test/java/org/apache/pinot/core/util/TableConfigUtilsTest.java:980]: Same here.

File: pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java (hunk 3 above)
Comment (icefury71) [reply pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java:503]: I suppose multi-value columns should be ok here - not really related to the encoding.

File: pinot-core/src/test/java/org/apache/pinot/core/util/TableConfigUtilsTest.java (hunk 4 above)
Comment (icefury71) [reply pinot-core/src/test/java/org/apache/pinot/core/util/TableConfigUtilsTest.java:962]: This is just for the unit test - I followed the same pattern as for others.

File: pinot-core/src/test/java/org/apache/pinot/core/util/TableConfigUtilsTest.java (hunk 5 above)
Comment (icefury71) [reply pinot-core/src/test/java/org/apache/pinot/core/util/TableConfigUtilsTest.java:980]: Same comment as above

**Review by mayankshriv — APPROVED** [review]

File: pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java (hunk 2 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java]: @icefury71  looks like this check was added back in another PR https://github.com/apache/incubator-pinot/pull/6620
Text index is supported on dictionary columns

Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java]: cc @jtao15

File: pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java [hunk 6]
```diff
@@ -488,18 +529,35 @@ private static void validateFieldConfigList(@Nullable List<FieldConfig> fieldCon
...
           "Column Name " + columnName + " defined in field config list must be a valid column defined in the schema");
 
-      if (fieldConfig.getEncodingType() == FieldConfig.EncodingType.DICTIONARY &&
-          indexingConfigs.getNoDictionaryColumns() != null) {
-        Preconditions.checkArgument(!indexingConfigs.getNoDictionaryColumns().contains(columnName),
-            "FieldConfig encoding type is different from indexingConfig for column: " + columnName);
+      List<String> noDictionaryColumns = indexingConfigs.getNoDictionaryColumns();
+      switch (fieldConfig.getEncodingType()) {
+        case RAW:
+          Preconditions.checkState(noDictionaryColumns != null && noDictionaryColumns.contains(columnName),
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java:539]: This check seems to be a needless one, given that we are in the case `RAW`. @icefury71 the new `FieldConfig` is where we want to move things, and eventually drop arrays in table config like `invertedIndexColumns`, and `noDictionaryColumns` etc.

File: pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java (hunk 2 above)
Comment (icefury71) [reply pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java]: @siddharthteotia are you sure it was #6620 ? It doesn't even touch TableConfigUtils

Comment (icefury71) [reply pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java]: @siddharthteotia the check doesn't exist as you can see : https://github.com/apache/incubator-pinot/blob/master/pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java#L550

**Pull request merged**